lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs

```diff
@@ -0,0 +1,232 @@
+            Records?.Enqueue(CreateMetricRecord(metric));
+        }
+
+        [JsonConstructor]
+        internal MetricTracker(IMetric metric, int changesSinceLastSink, ConcurrentQueue<MetricRecord> records, bool keepUpdateHistory)
+        {
+            Metric = metric;
+            Records = records;
+            KeepUpdateHistory = keepUpdateHistory;
+            ChangesSinceLastSink = changesSinceLastSink;
+        }
+
+        /// <summary>
+        /// Flush records currently held in the records queue.
+        /// </summary>
+        /// <returns>A queue containing all the flushed records.</returns>
+        internal IEnumerable<MetricRecord> FlushChangesSinceLastSink()
+        {
+            ConcurrentQueue<MetricRecord> records = new ConcurrentQueue<MetricRecord>();
+            if (Records != null)
+            {
+                while (Records.TryDequeue(out MetricRecord record))
+                {
+                    records.Enqueue(record);
+                }
+            }
+            else
+            {
+                // Records will be empty only on eval side when tracker doesn't keep history.
+                records.Enqueue(CreateMetricRecord(Metric));
+            }
+            Interlocked.Exchange(ref ChangesSinceLastSink, 0);
+            return records;
+        }
+
+        /// <summary>
+        /// When new metric data is received, update the value and records so it reflects the new data.
+        /// Called when Driver receives metrics from Evaluator.
+        /// </summary>
+        /// <param name="metric">Metric data received.</param>
+        internal MetricTracker UpdateMetric(MetricTracker metric)
+        {
+            if (metric.ChangesSinceLastSink > 0)
+            {
+                var recordsToAdd = metric.GetMetricRecords();
+                if (KeepUpdateHistory)
+                {
+                    foreach(MetricRecord record in recordsToAdd)
+                    {
+                        Records.Enqueue(record);
+                    }
+                }
+                else
+                {
+                    Interlocked.Exchange(ref Records, (ConcurrentQueue<MetricRecord>)recordsToAdd);
+                }
+            }
+            Interlocked.Add(ref ChangesSinceLastSink, metric.ChangesSinceLastSink);
+            return this;
+        }
+
+        /// <summary>
+        /// Get the metric with its most recent value.
+        /// </summary>
+        /// <returns></returns>
+        internal IMetric GetMetric()
+        {
+            return Metric;
+        }
+
+        /// <summary>
+        /// If KeepUpdateHistory is true, it will return all the records; otherwise, it will returen one record with the most recent value.
+        /// </summary>
+        /// <returns>The history of the metric records.</returns>
+        internal IEnumerable<MetricRecord> GetMetricRecords()
+        {
+            if (Records.IsEmpty)
```

Comment:
shouldn't we check for `KeepUpdateHistory` instead? I would write something like
```C#
if (KeepUpdateHistory)
{
    return Records;
}
return new[] { CreateMetricRecord(Metric) };
```

Reply:
The logic behind this is we only want to return CreateMetricRecord(Metric) when KeepUpdateHistory==false on the eval side. When eval passes metrics to the driver, there will always at least be one item in Records. The justification is if one were to call GetMetricRecords() from the driver, we want to return the last record that was passed from the evaluator, instead of the most current value of the metric itself.